docs: Add documentation for delete emoji endpoint.

Lalit3716 opened the pull request:
This endpoint was previously marked as `intentionally_undocumented` but that was a mistake.

Removed `intentionally_undocumented` and added proper documentation with valid `python_example` for this Endpoint.

<!-- Describe your pull request here.-->

Fixes: #24084
[CZO](https://chat.zulip.org/#narrow/stream/378-api-design/topic/Endpoint.20for.20delete.20emoji)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

![Screenshot 2023-01-25 130810](https://user-images.githubusercontent.com/84276404/214506136-e0180825-6b54-4b4f-9b26-b378a5841713.png)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

gnprice reviewed (changes requested):
Thanks @Lalit3716 for writing this up!

One thing I'd like to do before publishing documentation for this endpoint is to have a thread about it in `#api design` on chat.zulip.org, the same as we would do for a newly-added endpoint. That gives us the opportunity to discuss whether there's anything we'd like to tweak about this piece of API before we publish it.

(Put another way: I'd like to treat the existing state as being that this endpoint isn't in "the API", even though the server does respond to it.) 

I think it's quite likely that that thread will be short and we'll decide to publish the API exactly as it is. But I just want us to make sure we get to explicitly discuss that question.

Would you start a thread in `#api design` for that discussion?

Lalit3716 commented:
@gnprice Sure I opened the discussion thread and linked it in this PR description!

timabbott commented on zproject/urls.py:
```diff
@@ -260,7 +260,7 @@
-        DELETE=(delete_emoji, {"intentionally_undocumented"}),
+        DELETE=delete_emoji,
     ),
     # this endpoint throws a status code 400 JsonableError when it should be a 404.
```
You forgot to delete this comment.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8361,6 +8361,25 @@ paths:
+      tags: ["server_and_organizations"]
+      description: |
+        This endpoint is used to delete a custom emoji from the user's
+        organization.
```
We need a `**Changes**` entry here, with an API feature level bump and API changelog entry -- this is true for any API change, no matter how small.

https://zulip.readthedocs.io/en/latest/tutorials/new-feature-tutorial.html#update-documentation has details on how to do this correctly..

Also, the wording for the description feels a bit off in style compared with other endpoints; can you do an edit pass?

Lalit3716 replied:
What would you like in the description? I just followed `upload-custom-emoji` description style wording for it.

laurynmm replied:
I think it would be better to cut the first part and start with the verb, "Deactivate a custom emoji". I think it would make sense to make a similar change to the `api/upload-custom-emoji` documentation as well, but as a separate pull request.

Also, you'll want to use relative links for the help center in the API documentation (`/help/...`). And there's no linter rule, but in general it's good to keep the line breaks in this file similar to other sections.

gnprice reviewed (approved):
Thanks! Glad to have this documented.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8420,6 +8420,29 @@ paths:
+          example: green_tick
+      responses:
+        "200":
+          $ref: "#/components/responses/SimpleSuccess"
```
So the changes entry should be describing the API change -- which is just that we should document here a 404 response, and document as `**Changes**` that the error used to be a 400 status code.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8420,6 +8420,29 @@ paths:
+
+        Note: Deleted emoji will still be visible in old messages.
+
+        **Changes**: New in Zulip 7.0 (feature level 162).
```
This line is wrong -- the API endpoint has existed since before feature level 1. Please delete it.

Lalit3716 replied:
Oh, I was thinking from the user's perspective as it will look like it is a newly added API endpoint 😅, nevermind removed it!

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8420,6 +8420,29 @@ paths:
           $ref: "#/components/responses/SimpleSuccess"
+    delete:
+      operationId: delete-custom-emoji
+      summary: Delete custom emoji
```
Let's change the naming to be `deactivate-custom-emoji` -- that's how we explain the feature in the UI. https://zulip.com/help/custom-emoji#deactivate-custom-emoji. Similarly the text should all talk about "deactivate a custom emoji", etc.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8420,6 +8420,29 @@ paths:
+      summary: Delete custom emoji
+      tags: ["server_and_organizations"]
+      description: |
+        This endpoint is used to delete a custom emoji from the user's
```
"deactivate a custom emoji" here should be a Markdown link to the /help/ article for custom emoji noted above.

timabbott commented on api_docs/changelog.md:
```diff
@@ -674,7 +679,7 @@ No changes; feature level used for Zulip 5.0 release.
 **Feature level 83**
 
-* * [`POST /register`](/api/register-queue): The `cross_realm_bots`
+  * * [`POST /register`](/api/register-queue): The `cross_realm_bots`
```
This change looks like a typo.

Lalit3716 replied:
Right, it seems like it was my formatter messing with markdown files, fixed it!

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,11 @@ format used by the Zulip server that they are interacting with.
+**Feature level 164**
+
+* [`DELETE /realm/emoji/{emoji_name}`](/api/delete-custom-emoji): New API endpoint to delete
+  custom emoji from user's organization.
```
This is wrong -- the change is that the status code went from 400 to 404; the API endpoint' s behavior has otherwise been the same since feature level 1; it just wasn't documented before. We don't document the history of the API docs here; just the history of the API itself.

Lalit3716 replied:
Thanks for clearing that out!

Lalit3716 commented:
@timabbott Addressed your review once again!

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -8420,6 +8420,43 @@ paths:
+          required: true
+          in: path
+          description: |
+            The name of the custom emoji to delete.
```
nit: verb here should match the new name of the endpoint

Lalit3716 commented:
@gnprice Thanks for noticing that small detail fixed it!

Lalit3716 commented:
@timabbott Bumping this PR for a quick review since it's been some time.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,10 @@ format used by the Zulip server that they are interacting with.
 
+**Feature level 190**
+
+* [`DELETE /realm/emoji/{emoji_name}`](/api/deactivate-custom-emoji): Sending `404` status code when emoji does not exist instead of `400`.
```
Ths **Changes** entry both here and in `zulip.yaml` do not use the same phrasing style that we use for other API changes. @laurynmm can help if needed but you should be able to fix this by just studying the wording we use elsewhere.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -9128,6 +9128,43 @@ paths:
+
+        Note: Deactivated emoji will still be visible in old messages.
+
+        **Changes**: Starting with Zulip 8.0 (feature level 190), this endpoint will now return `404` status code when the emoji does not exist instead of `400`.
```
This should use the present tense when describing the current API, so "will now return" is wrong.

gnprice replied:
Here's a version:

```suggestion
        **Changes**: Before Zulip 8.0 (feature level 190), this endpoint returned an HTTP status code of 400 when the emoji did not exist, instead of 404.
```

The current API is already described by what's outside the "Changes" note (if it weren't, that'd be an important thing to fix), so the "Changes" note can focus on what used to be different.

gnprice commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,10 @@ format used by the Zulip server that they are interacting with.
 
+**Feature level 190**
+
+* [`DELETE /realm/emoji/{emoji_name}`](/api/deactivate-custom-emoji): This endpoint now returns `404` status code when an emoji does not exist instead of `400`.
```
```suggestion
* [`DELETE /realm/emoji/{emoji_name}`](/api/deactivate-custom-emoji): This endpoint
  now returns an HTTP status code of 404 when an emoji does not exist, instead of 400.
```

laurynmm replied:
Also, note the line breaks in this file for other feature level entries.

laurynmm commented on zerver/openapi/zulip.yaml:
```diff
@@ -9128,6 +9128,43 @@ paths:
+        This endpoint is used to [deactivate a custom emoji](https://zulip.com/help/custom-emoji#deactivate-custom-emoji) from the user's
+        organization.
+
+        Note: Deactivated emoji will still be visible in old messages.
```
Let's just make this a sentence, "Note that deactivated emoji ...".

Also, what happens to reactions, stream descriptions or user statuses that have the custom emoji when it's deactivated? Do they also remain like existing messages? It would be good to clarify that here.

Also, we should add a sentence, maybe before this about who can deactivate custom emoji. I believe it's admins can deactivate any custom emoji and other users can only deactivate custom emoji that they uploaded.

Lalit3716 replied:
Yeah so the deactivated emoji remains visible at all places so extended that sentence to include all other things.
Also, added the sentence for clarifying who can deactivate the custom emoji.

laurynmm reviewed (commented):
@Lalit3716 - Thanks for working on this! I agree with @gnprice's feedback and added a few additional comments. Let me know if you have any questions.

Just a note for when you've made the updates from the review feedback, I think it's clearer to post updated screenshots in a new comment for documentation PRs (instead of updating the original comment with screenshots). It's often clearer for folks who are reviewing the changes that way.

Lalit3716 commented:
Updates:
![Screenshot 2023-06-28 145703](https://github.com/zulip/zulip/assets/84276404/37082a4d-d8a2-4bc9-9a7d-5d81ef109c39)

- Added appropriate line breaks.
- Added a sentence for clarifying who can deactivate an emoji.
- Made changes to the sentences as suggested by @gnprice .

Ready for another look @timabbott @laurynmm @gnprice

timabbott commented:
Merged, thanks @Lalit3716!
